File: one_to_five_h/39-combination-sum.py
"""
    Leetcode 39
    From https://leetcode-cn.com/problems/combination-sum/
"""


def combinationSum(self, candidates, target):
    if not candidates:
        return []
    res = []
    candidates.sort()

    def dfs(start, sums, path):
        if sums == target:
            res.append(path)
            return
        if sums > target:
            return
        for i in range(start, len(candidates)):
            if sums + candidates[i] > target:
                break
            dfs(i, sums + candidates[i], path + [candidates[i]])
    dfs(0, 0, [])
    return res

File: backtrack/39-combination-sum.go
package main

func combinationSum(candidates []int, target int) [][]int {
	res := [][]int{}
	var dfs func([]int, int, int)
	dfs = func(path []int, sum int, start int) {
		if sum == target {
			tmp := make([]int, len(path))
			copy(tmp, path)
			res = append(res, tmp)
			return
		}
		if sum > target {
			return
		}
		for i := start; i < len(candidates); i++ {
			path = append(path, candidates[i])
			dfs(path, sum+candidates[i], i)
			path = path[:len(path)-1]
		}
	}
	dfs([]int{}, 0, 0)
	return res
}
